runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDos.java

```diff
@@ -0,0 +1,348 @@
+  }
+
+  private static SdkFunctionSpec toProto(DoFn<?, ?> fn, TupleTag<?> tag) {
+    return SdkFunctionSpec.newBuilder()
+        .setSpec(
+            FunctionSpec.newBuilder()
+                .setUrn(CUSTOM_JAVA_DO_FN_URN)
+                .setParameter(
+                    Any.pack(
+                        BytesValue.newBuilder()
+                            .setValue(
+                                ByteString.copyFrom(
+                                    SerializableUtils.serializeToByteArray(
+                                        DoFnAndMainOutput.of(fn, tag))))
+                            .build())))
+        .build();
+  }
+
+  private static DoFnAndMainOutput doFnAndMainOutputTagFromProto(SdkFunctionSpec fnSpec)
+      throws InvalidProtocolBufferException {
+    checkArgument(fnSpec.getSpec().getUrn().equals(CUSTOM_JAVA_DO_FN_URN));
+    byte[] serializedFn =
+        fnSpec.getSpec().getParameter().unpack(BytesValue.class).getValue().toByteArray();
+    return (DoFnAndMainOutput)
+        SerializableUtils.deserializeFromByteArray(serializedFn, "Custom DoFn And Main Output tag");
+  }
+
+  private static Optional<RunnerApi.Parameter> toProto(Parameter parameter) {
+    Cases<Type> runnerApiCases =
+        new Cases<Type>() {
+          @Override
+          public Type dispatch(StartBundleContextParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(FinishBundleContextParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(ProcessContextParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(OnTimerContextParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(WindowParameter p) {
+            return Type.WINDOW;
+          }
+
+          @Override
+          public Type dispatch(RestrictionTrackerParameter p) {
+            return Type.RESTRICTION_TRACKER;
+          }
+
+          @Override
+          public Type dispatch(StateParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(TimerParameter p) {
+            return null;
+          }
+        };
+    Type parameterType = parameter.match(runnerApiCases);
+    if (parameterType == null) {
+      return Optional.absent();
+    }
+    return Optional.of(RunnerApi.Parameter.newBuilder().setType(parameterType).build());
+  }
+
+  private static SideInput toProto(PCollectionView<?> view) {
+    Builder builder = SideInput.newBuilder();
+    builder.setAccessPattern(
+        FunctionSpec.newBuilder()
+            .setUrn(view.getViewFn().getMaterialization().getUrn())
+            .build());
+    builder.setViewFn(toProto(view.getViewFn()));
+    builder.setWindowMappingFn(toProto(view.getWindowMappingFn()));
+    return builder.build();
+  }
+
+  public static PCollectionView<?> fromProto(
```

Comment:
It's not a great scene. This one ends up needing a bunch of context because things like the `PCollection` are inputs of the `PTransform` and are only referenced by (two-indirection name) within the SideInput.

Some other things are `fooFromProto`, which I'm not opposed to.

Also reordered the parameters to make the conversion target first, then context